File: OsmAnd/src/net/osmand/plus/views/controls/maphudbuttons/ButtonPositionSize.java
package net.osmand.plus.views.controls.maphudbuttons;


import net.osmand.PlatformUtil;
import net.osmand.data.QuadRect;

import org.apache.commons.logging.Log;

import java.util.List;
import java.util.Locale;

public class ButtonPositionSize {

	private static final Log LOG = PlatformUtil.getLog(ButtonPositionSize.class);

	public static final int CELL_SIZE_DP = 8;
	public static final int DEF_MARGIN_DP = 4;

	public static final int MOVE_DESCENDANTS_ANY = 0;
	public static final int MOVE_DESCENDANTS_VERTICAL = 1;
	public static final int MOVE_DESCENDANTS_HORIZONTAL = 2;

	public static final int POS_FULL_WIDTH = 0;
	public static final int POS_LEFT = 1;
	public static final int POS_RIGHT = 2;

	public static final int POS_FULL_HEIGHT = 0;
	public static final int POS_TOP = 1;
	public static final int POS_BOTTOM = 2;

	private static final int MAX_MARGIN_BITS = 10;
	private static final int MAX_SIZE_BITS = 6;
	private static final int MARGIN_MASK = (1 << MAX_MARGIN_BITS) - 1;
	private static final int SIZE_MASK = (1 << MAX_SIZE_BITS) - 1;

	public int posH = POS_LEFT, posV = POS_TOP; // left, right, bottom, top
	public int marginX = 0, marginY = 0; // in 8dp scale
	public int width = 7, height = 7; // in 8dp scale including shadow
	public boolean xMove = false, yMove = false;
	public int moveDescendants = MOVE_DESCENDANTS_ANY;
	public String id;

	private final QuadRect bounds = new QuadRect();

	public ButtonPositionSize(String id) {
		this(id, 7, true, true);
	}

	public ButtonPositionSize(String id, int sz8dp, boolean left, boolean top) {
		this.id = id;
		width = height = sz8dp;
		this.posH = left ? POS_LEFT : POS_RIGHT;
		this.posV = top ? POS_TOP : POS_BOTTOM;
	}

	public ButtonPositionSize(String id, int sz8dp, int posH, int posV) {
		this.id = id;
		width = height = sz8dp;
		this.posH = posH;
		this.posV = posV;
		validate();
	}

	public ButtonPositionSize setMoveVertical() {
		this.yMove = true;
		return this;
	}

	public ButtonPositionSize setMoveDescendantsAny() {
		this.moveDescendants = MOVE_DESCENDANTS_ANY;
		return this;
	}

	public ButtonPositionSize setMoveDescendantsVertical() {
		this.moveDescendants = MOVE_DESCENDANTS_VERTICAL;
		return this;
	}

	public ButtonPositionSize setMoveDescendantsHorizontal() {
		this.moveDescendants = MOVE_DESCENDANTS_HORIZONTAL;
		return this;
	}

	public ButtonPositionSize setMoveHorizontal() {
		this.xMove = true;
		return this;
	}

	public ButtonPositionSize setSize(int width8dp, int height8dp) {
		this.width = width8dp;
		this.height = height8dp;
		return this;
	}

	public ButtonPositionSize setPositionHorizontal(int posH) {
		this.posH = posH;
		validate();
		return this;
	}

	public ButtonPositionSize setPositionVertical(int posV) {
		this.posV = posV;
		validate();
		return this;
	}

	public ButtonPositionSize setMargin(int marginX, int marginY) {
		this.marginX = marginX;
		this.marginY = marginY;
		return this;
	}

	public boolean isLeft() {
		return posH == POS_LEFT;
	}

	public boolean isRight() {
		return posH == POS_RIGHT;
	}

	public boolean isTop() {
		return posV == POS_TOP;
	}

	public boolean isBottom() {
		return posV == POS_BOTTOM;
	}

	public boolean isFullWidth() {
		return posH == POS_FULL_WIDTH;
	}

	public boolean isFullHeight() {
		return posV == POS_FULL_HEIGHT;
	}

	public long toLongValue() {
		long vl = 0;
		vl = (vl << 2) + moveDescendants;
		vl = (vl << 2) + posV;
		vl = (vl << 1) + (yMove ? 1 : 0);
		vl = (vl << MAX_MARGIN_BITS) + Math.min(marginY, MARGIN_MASK);
		vl = (vl << MAX_SIZE_BITS) + Math.min(height, SIZE_MASK);
		vl = (vl << 2) + posH;
		vl = (vl << 1) + (xMove ? 1 : 0);
		vl = (vl << MAX_MARGIN_BITS) + Math.min(marginX, MARGIN_MASK);
		vl = (vl << MAX_SIZE_BITS) + Math.min(width, SIZE_MASK);
		return vl;
	}

	public void calcGridPositionFromPixel(float dpToPix, int widthPx, int heightPx,
	                                      boolean gravLeft, int x, boolean gravTop, int y) {
		float calc;
		if (x < widthPx / 2) {
			this.posH = gravLeft ? POS_LEFT : POS_RIGHT;
			calc = x / dpToPix;
		} else {
			this.posH = gravLeft ? POS_RIGHT : POS_LEFT;
			calc = (widthPx - x) / dpToPix - this.width * CELL_SIZE_DP;
		}
		this.marginX = Math.max(0, Math.round((calc - DEF_MARGIN_DP) / CELL_SIZE_DP));
		if (y < heightPx / 2) {
			this.posV = gravTop ? POS_TOP : POS_RIGHT;
			calc = y / dpToPix;
		} else {
			this.posV = gravTop ? POS_RIGHT : POS_TOP;
			calc = (heightPx - y) / dpToPix - this.height * CELL_SIZE_DP;
		}
		this.marginY = Math.max(0, Math.round((calc - DEF_MARGIN_DP) / CELL_SIZE_DP));
	}

	public int getYStartPix(float dpToPix) {
		return (int) ((marginY * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix);
	}

	public int getYEndPix(float dpToPix) {
		return (int) (((marginY + width) * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix);
	}

	public int getXStartPix(float dpToPix) {
		return (int) ((marginX * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix);
	}

	public int getXEndPix(float dpToPix) {
		return (int) (((marginX + width) * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix);
	}

	public int getWidthPix(float dpToPix) {
		return (int) ((width * CELL_SIZE_DP) * dpToPix);
	}

	public ButtonPositionSize fromLongValue(long v) {
		width = (int) (v & SIZE_MASK);
		v = v >> MAX_SIZE_BITS;
		marginX = (int) (v & MARGIN_MASK);
		v = v >> MAX_MARGIN_BITS;
		xMove = v % 2 == 1;
		v = v >> 1;
		posH = (int) (v % 4);
		v = v >> 2;
		height = (int) (v & SIZE_MASK);
		v = v >> MAX_SIZE_BITS;
		marginY = (int) (v & MARGIN_MASK);
		v = v >> MAX_MARGIN_BITS;
		yMove = v % 2 == 1;
		v = v >> 1;
		posV = (int) (v % 4);
		v = v >> 2;
		moveDescendants = (int) (v % 4);
		v = v >> 2;
		validate();
		return this;
	}

	private void validate() {
		if (posH == POS_FULL_WIDTH && posV == POS_FULL_HEIGHT) {
			System.err.println("Error parsing " + this + " as full width + full height");
			posH = POS_RIGHT;
		}
	}

	@Override
	public String toString() {
		return String.format(Locale.US, "Pos %10s x=(%s->%d%s), y=(%s->%d%s), w=%2d, h=%2d", id,
				posH == POS_FULL_WIDTH ? "full_w" : posH == POS_LEFT ? "left " : "right", marginX, xMove ? "+" : " ",
				posV == POS_FULL_HEIGHT ? "full_h" : posV == POS_TOP ? "top " : "bott", marginY, yMove ? "+" : " ", width, height);
	}

	public boolean overlap(ButtonPositionSize position) {
		return QuadRect.intersects(this.bounds, position.bounds);
	}

	public static boolean computeNonOverlap(int space, List<ButtonPositionSize> buttons, int totalWidth, int totalHeight) {
		int MAX_ITERATIONS = 1000, iter = 0;
		for (ButtonPositionSize button : buttons)
			button.updateBounds(totalWidth, totalHeight);

		for (int fixedPos = buttons.size() - 1; fixedPos >= 0; ) {
			if (iter++ > MAX_ITERATIONS) {
				LOG.error("Relayout is broken");
				return false;
			}
			boolean overlap = false;
			ButtonPositionSize button = buttons.get(fixedPos);
			for (int i = fixedPos + 1; i < buttons.size(); i++) {
				ButtonPositionSize check = buttons.get(i);
				if (button.overlap(check)) {
					overlap = true;
					if (moveButton(space, check, button, totalWidth, totalHeight)) {
						check.updateBounds(totalWidth, totalHeight);
					}
					fixedPos = i;
					break;
				}
			}
			if (!overlap) {
				fixedPos--;
			}
		}
		return true;
	}

	private void updateBounds(int totalWidth, int totalHeight) {
		double left, right, top, bottom;

		if (posH == POS_FULL_WIDTH) {
			left = 0;
			right = totalWidth;
		} else if (posH == POS_LEFT) {
			left = marginX;
			right = left + width;
		} else {
			right = totalWidth - marginX;
			left = right - width;
		}

		if (posV == POS_FULL_HEIGHT) {
			top = 0;
			bottom = totalHeight;
		} else if (posV == POS_TOP) {
			top = marginY;
			bottom = top + height;
		} else {
			bottom = totalHeight - marginY;
			top = bottom - height;
		}
		bounds.left = left;
		bounds.right = right;
		bounds.top = top;
		bounds.bottom = bottom;
	}

	private static boolean moveButton(int space, ButtonPositionSize toMove,
			ButtonPositionSize overlap,	int totalWidth, int totalHeight) {
		boolean xMove = false;
		boolean yMove = false;
		if (overlap.moveDescendants == MOVE_DESCENDANTS_ANY) {
			if (overlap.posH == POS_FULL_WIDTH) {
				yMove = true;
			} else if (overlap.posV == POS_FULL_HEIGHT) {
				xMove = true;
			} else {
				if (toMove.xMove) {
					xMove = toMove.xMove;
				}
				if (toMove.yMove) {
					yMove = toMove.yMove;
				}
			}
		} else if (overlap.moveDescendants == MOVE_DESCENDANTS_VERTICAL) {
			yMove = true;
		} else if (overlap.moveDescendants == MOVE_DESCENDANTS_HORIZONTAL) {
			xMove = true;
		}

		int newX = xMove ? space + overlap.marginX + overlap.width : toMove.marginX;
		int newY = yMove ? space + overlap.marginY + overlap.height : toMove.marginY;

		if (newX + toMove.width > totalWidth) {
			newX = Math.max(0, totalWidth - toMove.width);
			if (!yMove) {
				int desired = space + overlap.marginY + overlap.height;
				newY = Math.max(0, Math.min(desired, totalHeight - toMove.height));
			}
		}
		if (newY + toMove.height > totalHeight) {
			newY = Math.max(0, totalHeight - toMove.height);
			if (!xMove) {
				int desired = space + overlap.marginX + overlap.width;
				newX = Math.max(0, Math.min(desired, totalWidth - toMove.width));
			}
		}
		boolean moved = (newX != toMove.marginX) || (newY != toMove.marginY);
		toMove.marginX = newX;
		toMove.marginY = newY;

		return moved;
	}
}

File: OsmAnd-shared/src/commonMain/kotlin/net/osmand/shared/grid/ButtonPositionSize.kt
package net.osmand.shared.grid

import net.osmand.shared.data.KQuadRect
import net.osmand.shared.util.LoggerFactory

import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt

class ButtonPositionSize {

	val id: String

	var posH: Int = POS_LEFT
	var posV: Int = POS_TOP
	var marginX: Int = 0
	var marginY: Int = 0
	var width: Int = 7
	var height: Int = 7
	var xMove: Boolean = false
	var yMove: Boolean = false
	var moveDescendants: Int = MOVE_DESCENDANTS_ANY

	internal val bounds = KQuadRect()

	constructor(id: String) : this(id, 7, true, true)

	constructor(id: String, sz8dp: Int, left: Boolean, top: Boolean) {
		this.id = id
		width = sz8dp
		height = sz8dp
		this.posH = if (left) POS_LEFT else POS_RIGHT
		this.posV = if (top) POS_TOP else POS_BOTTOM
	}

	constructor(id: String, sz8dp: Int, posH: Int, posV: Int) {
		this.id = id
		width = sz8dp
		height = sz8dp
		this.posH = posH
		this.posV = posV
		validate()
	}

	fun setMoveVertical() = apply { this.yMove = true }

	fun setMoveDescendantsAny() = apply { this.moveDescendants = MOVE_DESCENDANTS_ANY }

	fun setMoveDescendantsVertical() = apply { this.moveDescendants = MOVE_DESCENDANTS_VERTICAL }

	fun setMoveDescendantsHorizontal() = apply { this.moveDescendants = MOVE_DESCENDANTS_HORIZONTAL }

	fun setMoveHorizontal() = apply { this.xMove = true }

	fun setNonMoveable() = apply { this.xMove = false; this.yMove = false; }


	fun setSize(width8dp: Int, height8dp: Int) = apply {
		this.width = width8dp
		this.height = height8dp
	}

	fun setPositionHorizontal(posH: Int) = apply {
		this.posH = posH
		validate()
	}

	fun setPositionVertical(posV: Int) = apply {
		this.posV = posV
		validate()
	}

	fun setMargin(marginX: Int, marginY: Int) = apply {
		this.marginX = marginX
		this.marginY = marginY
	}

	val isLeft: Boolean get() = posH == POS_LEFT
	val isRight: Boolean get() = posH == POS_RIGHT
	val isTop: Boolean get() = posV == POS_TOP
	val isBottom: Boolean get() = posV == POS_BOTTOM
	val isFullWidth: Boolean get() = posH == POS_FULL_WIDTH
	val isFullHeight: Boolean get() = posV == POS_FULL_HEIGHT

	fun toLongValue(): Long {
		var vl = 0L
		vl = (vl shl 2) + moveDescendants
		vl = (vl shl 2) + posV
		vl = (vl shl 1) + if (yMove) 1 else 0
		vl = (vl shl MAX_MARGIN_BITS) + min(marginY, MARGIN_MASK)
		vl = (vl shl MAX_SIZE_BITS) + min(height, SIZE_MASK)
		vl = (vl shl 2) + posH
		vl = (vl shl 1) + if (xMove) 1 else 0
		vl = (vl shl MAX_MARGIN_BITS) + min(marginX, MARGIN_MASK)
		vl = (vl shl MAX_SIZE_BITS) + min(width, SIZE_MASK)
		return vl
	}

	fun calcGridPositionFromPixel(
		dpToPix: Float, widthPx: Int, heightPx: Int,
		gravLeft: Boolean, x: Int, gravTop: Boolean, y: Int
	) {
		val calcX: Float
		if (x < widthPx / 2) {
			this.posH = if (gravLeft) POS_LEFT else POS_RIGHT
			calcX = x / dpToPix
		} else {
			this.posH = if (gravLeft) POS_RIGHT else POS_LEFT
			calcX = (widthPx - x) / dpToPix - this.width * CELL_SIZE_DP
		}
		this.marginX = max(0, ((calcX - DEF_MARGIN_DP) / CELL_SIZE_DP).roundToInt())

		val calcY: Float
		if (y < heightPx / 2) {
			this.posV = if (gravTop) POS_TOP else POS_BOTTOM
			calcY = y / dpToPix
		} else {
			this.posV = if (gravTop) POS_BOTTOM else POS_TOP
			calcY = (heightPx - y) / dpToPix - this.height * CELL_SIZE_DP
		}
		this.marginY = max(0, ((calcY - DEF_MARGIN_DP) / CELL_SIZE_DP).roundToInt())
	}

	fun getYStartPix(dpToPix: Float): Int {
		return ((marginY * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix).toInt()
	}

	fun getYEndPix(dpToPix: Float): Int {
		return (((marginY + height) * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix).toInt()
	}

	fun getXStartPix(dpToPix: Float): Int {
		return ((marginX * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix).toInt()
	}

	fun getXEndPix(dpToPix: Float): Int {
		return (((marginX + width) * CELL_SIZE_DP + DEF_MARGIN_DP) * dpToPix).toInt()
	}

	fun getWidthPix(dpToPix: Float): Int {
		return ((width * CELL_SIZE_DP) * dpToPix).toInt()
	}

	fun fromLongValue(v: Long) = apply {
		var value = v
		width = (value and SIZE_MASK.toLong()).toInt()
		value = value shr MAX_SIZE_BITS
		marginX = (value and MARGIN_MASK.toLong()).toInt()
		value = value shr MAX_MARGIN_BITS
		xMove = value % 2 == 1L
		value = value shr 1
		posH = (value % 4).toInt()
		value = value shr 2
		height = (value and SIZE_MASK.toLong()).toInt()
		value = value shr MAX_SIZE_BITS
		marginY = (value and MARGIN_MASK.toLong()).toInt()
		value = value shr MAX_MARGIN_BITS
		yMove = value % 2 == 1L
		value = value shr 1
		posV = (value % 4).toInt()
		value = value shr 2
		moveDescendants = (value % 4).toInt()
		validate()
	}

	private fun validate() {
		if (posH == POS_FULL_WIDTH && posV == POS_FULL_HEIGHT) {
			LOG.error("Error parsing $this as full width + full height")
			posH = POS_RIGHT
		}
	}

	override fun toString(): String {
		val posHStr = when (posH) {
			POS_FULL_WIDTH -> "full_w"
			POS_LEFT -> "left "
			else -> "right"
		}
		val posVStr = when (posV) {
			POS_FULL_HEIGHT -> "full_h"
			POS_TOP -> "top "
			else -> "bott"
		}
		val xMoveIndicator = if (xMove) "+" else " "
		val yMoveIndicator = if (yMove) "+" else " "
		val paddedWidth = width.toString().padStart(2)
		val paddedHeight = height.toString().padStart(2)
		return "Pos ${id.padStart(10)} x=($posHStr->${marginX}$xMoveIndicator), y=($posVStr->${marginY}$yMoveIndicator), w=$paddedWidth, h=$paddedHeight"
	}

	fun overlap(position: ButtonPositionSize): Boolean {
		return KQuadRect.intersects(this.bounds, position.bounds)
	}

	private fun updateBounds(totalWidth: Int, totalHeight: Int) {
		val left: Double
		val right: Double
		val top: Double
		val bottom: Double

		if (posH == POS_FULL_WIDTH) {
			left = 0.0
			right = totalWidth.toDouble()
		} else if (posH == POS_LEFT) {
			left = marginX.toDouble()
			right = left + width
		} else {
			right = (totalWidth - marginX).toDouble()
			left = right - width
		}

		if (posV == POS_FULL_HEIGHT) {
			top = 0.0
			bottom = totalHeight.toDouble()
		} else if (posV == POS_TOP) {
			top = marginY.toDouble()
			bottom = top + height
		} else {
			bottom = (totalHeight - marginY).toDouble()
			top = bottom - height
		}
		bounds.left = left
		bounds.right = right
		bounds.top = top
		bounds.bottom = bottom
	}

	companion object {

		private val LOG = LoggerFactory.getLogger("ButtonPositionSize")

		private const val MAX_ITERATIONS = 1000
		private const val MAX_STUCK_ATTEMPTS = 20

		var DEBUG_PRINT: Boolean = true; // false
		var DEBUG_PRINT_MOVE: Boolean = true; // false
		var SIMPLIFIED_ALGORITHM: Boolean = true; // original algorithm is more complex but produces different result with broken layout

		const val CELL_SIZE_DP = 8
		const val DEF_MARGIN_DP = 4

		const val MOVE_DESCENDANTS_ANY = 0
		const val MOVE_DESCENDANTS_VERTICAL = 1
		const val MOVE_DESCENDANTS_HORIZONTAL = 2

		const val POS_FULL_WIDTH = 0
		const val POS_LEFT = 1
		const val POS_RIGHT = 2

		const val POS_FULL_HEIGHT = 0
		const val POS_TOP = 1
		const val POS_BOTTOM = 2

		private const val MAX_MARGIN_BITS = 10
		private const val MAX_SIZE_BITS = 6
		private const val MARGIN_MASK = (1 shl MAX_MARGIN_BITS) - 1
		private const val SIZE_MASK = (1 shl MAX_SIZE_BITS) - 1

		fun computeNonOverlap(
			space: Int,
			buttons: List<ButtonPositionSize>,
			totalWidth: Int,
			totalHeight: Int
		): Boolean {
			buttons.forEach {
				it.updateBounds(totalWidth, totalHeight);
			}
			if (DEBUG_PRINT) {
				LOG.info("---- Layout w=$totalWidth h=$totalHeight spc=$space");
				buttons.forEach {
					LOG.info("Before Button $it ${it.bounds}");
				}
			}
			var iter = 0

			var fixedPos = buttons.size - 1
			// we tdo it in forward iteration (I step all [1, .. I-1] are arranged)
			// not(SIMPLIFIED_ALGORITHM) is original algorithm it's more complex but produces different result with broken layout
			if (SIMPLIFIED_ALGORITHM) {
				var i = 1;
				while (i < buttons.size) {
					if (iter++ > MAX_ITERATIONS) {
						LOG.error("Relayout is broken.")
						return false
					}
					var movedOnce = false;
					var failMove = false;
					val button = buttons[i]
					var ox = button.marginX;
					var oy = button.marginY;
					for (j in i - 1 downTo 0) {
						val check = buttons[j]
						if (button.overlap(check)) {
							val moved = moveButton(space, button, check, totalWidth, totalHeight)
							if (!moved) {
								failMove = true;
								break;
							}
							if (check.xMove || check.yMove) {
								// if overlapped button / panel moveable save it
								// do not overlap buttons but allow to overlap buttons over pannels
								ox = button.marginX;
								oy = button.marginY;

							}
							movedOnce = true
							if (DEBUG_PRINT_MOVE) {
								LOG.info("Move $button because $check new bounds ${button.bounds}");
							}
						}
					}
					if (failMove) {
						button.marginX = ox;
						button.marginY = oy;
						button.updateBounds(totalWidth, totalHeight);
						i++
					} else if (!movedOnce) {
						i++
					}
				}
			} else {
				while (fixedPos >= 0) {
					if (iter++ > MAX_ITERATIONS) {
						LOG.error("Relayout is broken.")
						return false
					}
					var overlap = false
					val button = buttons[fixedPos]
					for (i in (fixedPos + 1) until buttons.size) {
						val check = buttons[i]
						if (button.overlap(check)) {
							val moved = moveButton(space, check, button, totalWidth, totalHeight)
							if (!moved) {
								// skip unmoveable situations
								continue;
							}
							overlap = true
							check.updateBounds(totalWidth, totalHeight)
							if (DEBUG_PRINT_MOVE) {
								LOG.info("Move $check because $button new bounds ${check.bounds}");
							}
							fixedPos = i
							break
						}
					}

					if (!overlap) {
						fixedPos--
					}
				}
			}
			if (DEBUG_PRINT) {
				buttons.forEach {
					LOG.info("After Button $it ${it.bounds}");
				}
				LOG.info("++++ Layout w=$totalWidth h=$totalHeight spc=$space");
			}
			return true
		}

		private fun moveAndCheck(toMove: ButtonPositionSize, overlap: ButtonPositionSize,
								 totalWidth: Int, totalHeight: Int, mx: Int, my: Int): Boolean {
			toMove.marginX += mx;
			toMove.marginY += my;
			toMove.updateBounds(totalWidth, totalHeight);
			return toMove.overlap(overlap);
		}
		private fun moveButton(space: Int, toMove: ButtonPositionSize, overlap: ButtonPositionSize,
							   totalWidth: Int, totalHeight: Int): Boolean {
			var xMove = false
			var yMove = false
			if (!toMove.xMove && !toMove.yMove) {
				return false
			}

			if (overlap.moveDescendants == MOVE_DESCENDANTS_ANY) {
				if (overlap.posH == POS_FULL_WIDTH) {
					yMove = true
				} else if (overlap.posV == POS_FULL_HEIGHT) {
					xMove = true
				} else {
					xMove = toMove.xMove
					yMove = toMove.yMove
				}
			} else if (overlap.moveDescendants == MOVE_DESCENDANTS_VERTICAL) {
				yMove = true
			} else if (overlap.moveDescendants == MOVE_DESCENDANTS_HORIZONTAL) {
				xMove = true
			}
			var ox = toMove.marginX;
			var oy = toMove.marginY;
			var ok = false;
			while (!ok && xMove && toMove.marginX <= totalWidth - toMove.width) {
				if (!moveAndCheck(toMove, overlap, totalWidth, totalHeight, 1, 0)) {
					ok = true;
				}
			}
			if (!ok) {
				toMove.marginX = ox;
				toMove.updateBounds(totalWidth, totalHeight);
			}
			while (!ok && yMove && toMove.marginY <= totalHeight - toMove.height) {
				if (!moveAndCheck(toMove, overlap, totalWidth, totalHeight, 0, 1)) {
					ok = true;
				}
			}
			if (!ok) {
				toMove.marginY = oy;
				toMove.updateBounds(totalWidth, totalHeight);
			}
			return ok;
		}
	}
}
